### PR TITLE
fix(github): replace invalid Repository.teams GraphQL field

### PR DESCRIPTION
**Pull request by onuryilmaz**
## Summary

- GitHub's GraphQL API does not expose a `teams` field on the `Repository` type, causing every `GithubOrganization` reconcile to fail with `Field 'teams' doesn't exist on type 'Repository'` (root cause of the nightly failures in run [#28521420326](https://github.com/cloudoperators/repo-guard/actions/runs/28521420326/job/84546321018))
- Rewrites `ExtendedListGraphQL` to traverse the graph in the correct direction: `organisation → teams → repositories`, building a `repoName → []teams` map first, then fetching repo metadata separately and combining
- For the rare edge case of a team with >100 repositories, a dedicated `teamReposQuery` paginates that team's repos in isolation (avoids the cursor-scope bug of the previous approach)

## Test plan

- [ ] All existing unit tests updated to match the new two-query pattern and pass (`go test ./internal/github/`)
- [ ] `TestExtendedListGraphQL_TeamsPagination` — verifies multiple pages of teams are merged correctly
- [ ] `TestExtendedListGraphQL_TeamRepoOverflow` — verifies the >100-repos single-team fallback query
- [ ] Nightly triggered on this branch: [run #28525623010](https://github.com/cloudoperators/repo-guard/actions/runs/28525623010)

### CODE REVIEW

File: internal/github/repos_graphql_test.go [hunk 1]
```diff
@@ -72,28 +90,35 @@ func buildOrgReposGraphQLResponse(nodes []map[string]any, hasNextPage bool, endC
...
+		"slug": slug,
+		"repositories": map[string]any{
 			"pageInfo": map[string]any{
-				"hasNextPage": teamsHasNextPage,
+				"hasNextPage": reposHasNextPage,
 				"endCursor":   "",
 			},
-			"edges": teams,
+			"edges": repoEdges,
 		},
```
[the remark this answers is not in the document]
Comment (onuryilmaz) [reply internal/github/repos_graphql_test.go:115]: Fixed in c977e76. Added an `endCursor` parameter to `buildTeamNode` and updated `TestExtendedListGraphQL_TeamRepoOverflow` to pass `"repo-cursor-overflow"` so `fetchRemainingTeamRepos` is called with a real cursor instead of an empty string.

File: internal/github/repos_graphql.go [hunk 2]
```diff
@@ -63,26 +100,127 @@ func graphqlPermissionToTeamPermission(p string) repoguardsapv1.GithubTeamPermis
...
+	return nil
+}
+
 // ExtendedListGraphQL fetches all organisation repositories and their team permissions
-// using a single paginated GraphQL query instead of N+1 REST calls.
-// For 300 repositories this reduces ~303 REST calls to 3–4 GraphQL calls per reconcile.
+// using GraphQL queries instead of N+1 REST calls.
+// For 300 repositories this significantly reduces REST API consumption per reconcile.
 //
 // Repos that are archived or disabled are excluded (same semantics as List/ExtendedList).
```
[the remark this answers is not in the document]
Comment (onuryilmaz) [reply internal/github/repos_graphql.go:226]: Fixed in c977e76. The mock's `/api/graphql` handler now inspects the request variables and dispatches to three response shapes: `teamCursor` → `organization.teams.nodes[].repositories` (orgTeamsWithReposQuery), `teamSlug` → `organization.team.repositories` (teamReposQuery), default → `organization.repositories.nodes` metadata only (orgReposQuery). Controller tests now converge correctly — confirmed by `make controller-test` passing locally.

File: internal/github/mock_server.go [hunk 3]
```diff
@@ -883,53 +883,127 @@ func registerMockHandlers(mux *http.ServeMux, cfg MockConfig) {
...
+				}
+				teamNodes = append(teamNodes, map[string]any{
+					"slug": t.Slug,
 					"repositories": map[string]any{
 						"pageInfo": map[string]any{"hasNextPage": false, "endCursor": ""},
-						"nodes":    nodes,
+						"edges":    edges,
+					},
+				})
+			}
```
[the remark this answers is not in the document]
Comment (onuryilmaz) [reply internal/github/mock_server.go:971]: Fixed in d190f95. After the `teamsSnap` loop, a second loop over `teamRepoEdges` now emits a team node for any slug not already in `seen`, covering team slugs that exist only in `teamRepoPerms` (dynamically created via REST but not seeded in `cfg.Teams`).

File: internal/github/repos_graphql_test.go [hunk 4]
```diff
@@ -232,67 +277,118 @@ func TestExtendedListGraphQL_ArchivedFiltered(t *testing.T) {
...
+	if err != nil {
+		t.Fatalf("ExtendedListGraphQL: unexpected error: %v", err)
+	}
+	if len(priv) != 1 {
+		t.Fatalf("expected 1 private repo, got %v", priv)
+	}
+	// repo-x should have both team-a (push) and team-b (pull) permissions.
+	if len(priv[0].Teams) != 2 {
+		t.Errorf("expected 2 teams on repo-x, got %v", priv[0].Teams)
+	}
```
[the remark this answers is not in the document]
Comment (onuryilmaz) [reply internal/github/repos_graphql_test.go:319]: Fixed in d190f95. Added order-independent assertions that check both slug names (`team-a`, `team-b`) and their mapped permissions (`push`, `pull`), so permission regressions or incorrect pagination merges will now cause the test to fail.

File: internal/github/repos_graphql.go [hunk 5]
```diff
@@ -13,30 +13,67 @@ import (
...
+
+// teamReposQuery is the GraphQL query struct for fetching repositories of a
+// single team. Used as a fallback when a team has more than 100 repositories.
+type teamReposQuery struct {
+	Organization struct {
+		Team struct {
+			Repositories struct {
+				PageInfo struct {
+					HasNextPage githubv4.Boolean
+					EndCursor   githubv4.String
```
[the remark this answers is not in the document]
Comment (onuryilmaz) [reply internal/github/repos_graphql.go]: Fixed in 2037afe. `teamReposQuery.Organization.Team` is now a `*struct` (pointer). `fetchRemainingTeamRepos` checks for nil immediately after the query and breaks out of the loop, treating a deleted/renamed team as an empty result rather than panicking or failing the decode.

File: internal/github/repos_graphql.go [hunk 6]
```diff
@@ -63,26 +100,127 @@ func graphqlPermissionToTeamPermission(p string) repoguardsapv1.GithubTeamPermis
...
+					Team:       slug,
+					Permission: perm,
+				})
+			}
+
+			// Rare edge case: team has >100 repos. Use a dedicated single-team
+			// query so the repoCursor applies only to this team.
+			if bool(team.Repositories.PageInfo.HasNextPage) {
+				if err := t.fetchRemainingTeamRepos(ctx, slug, team.Repositories.PageInfo.EndCursor, repoTeams); err != nil {
+					return nil, err
```
[the remark this answers is not in the document]
Comment (onuryilmaz) [reply internal/github/repos_graphql.go:153]: Fixed in 2037afe. The `HasNextPage` check now also requires `EndCursor != ""` before calling `fetchRemainingTeamRepos`, so a malformed response with `HasNextPage=true` and an empty cursor will be silently skipped rather than triggering an infinite loop on `after: ""`.

File: internal/github/repos_graphql.go [hunk 7]
```diff
@@ -63,26 +103,136 @@ func graphqlPermissionToTeamPermission(p string) repoguardsapv1.GithubTeamPermis
...
+				}
+			}
+		}
+
+		if !bool(query.Organization.Teams.PageInfo.HasNextPage) {
+			break
+		}
+		cursor := query.Organization.Teams.PageInfo.EndCursor
+		teamCursor = &cursor
```
[the remark this answers is not in the document]
Comment (onuryilmaz) [reply internal/github/repos_graphql.go:167]: Fixed in 109d84d. Added a `cursor == ""` guard in `buildRepoTeamsMap` after the `HasNextPage` check — breaks the loop instead of setting `teamCursor` to an empty string and repeating the same page.

File: internal/github/repos_graphql.go [hunk 8]
```diff
@@ -63,26 +103,136 @@ func graphqlPermissionToTeamPermission(p string) repoguardsapv1.GithubTeamPermis
...
+			repoTeams[repoName] = append(repoTeams[repoName], repoguardsapv1.GithubTeamWithPermission{
+				Team:       slug,
+				Permission: perm,
+			})
+		}
+
+		if !bool(query.Organization.Team.Repositories.PageInfo.HasNextPage) {
+			break
+		}
+		cursor = query.Organization.Team.Repositories.PageInfo.EndCursor
```
[the remark this answers is not in the document]
Comment (onuryilmaz) [reply internal/github/repos_graphql.go]: Fixed in 109d84d. Added the same empty-cursor guard in `fetchRemainingTeamRepos`: reads into a `nextCursor` variable, breaks if it is empty, and only assigns to `cursor` when it is non-empty — preventing duplicate result accumulation.

File: internal/github/repos_graphql_test.go [hunk 9]
```diff
@@ -232,67 +277,129 @@ func TestExtendedListGraphQL_ArchivedFiltered(t *testing.T) {
...
 	}
-	// Teams should come from REST fallback, not GraphQL partial result.
-	if len(priv[0].Teams) != 1 || priv[0].Teams[0].Team != "team-rest" {
-		t.Errorf("expected REST-sourced team, got %v", priv[0].Teams)
+	// Both repos should be associated with big-team.
+	for _, repo := range priv {
+		if len(repo.Teams) != 1 || repo.Teams[0].Team != "big-team" {
+			t.Errorf("repo %q: expected big-team, got %v", repo.Name, repo.Teams)
+		}
 	}
```
[the remark this answers is not in the document]
Comment (onuryilmaz) [reply internal/github/repos_graphql_test.go:404]: Fixed in 8d5d61d. Added per-repo permission assertions: `repo-1` (WRITE→`push`) and `repo-2` (ADMIN→`admin`) are now explicitly checked, so any bug that drops or mis-maps permissions during the first-page/overflow merge will be caught.
